Show realm migration progress via a simple UI

peppy opened the pull request:
This is working and ready to go, but needs a touch more consideration with regards to subscriptions. Will be addressing that separately and blocking this for now.

- Closes https://github.com/ppy/osu/issues/16554

peppy committed:
Expose EF context factory for use in external migration logic

peppy committed:
Show realm migration progress at `Loader`

peppy commented:
Unblocking, added dependency on subscription fixes PR.

frenzibyte commented on osu.Game/Database/EFToRealmMigrator.cs:
```diff
@@ -4,52 +4,130 @@
         {
+            base.LoadComplete();
+
+            // needs to be run on the update thread because of realm BlockAllOperations.
+            // maybe we can work around this? not sure..
```
Is this comment still valid? this is invoked in `LoadComplete` which should guarantee running from update thread.

peppy committed:
Create backup before any realm contexts are used

peppy committed:
Remove using statements

peppy commented:
In the interest of keeping things moving forward as I have a lot more realm changes incoming, I've fixed this to not require the subscription changes. As pointed out by my own comment, the only reason we're facing issues in the first place is due to taking the backup. So I've moved that backup to `OsuGameBase` to have it run before other realm usages.

Note that this means running visual tests will take a backup but not migrate. I don't see that as a huge issue though - it will resolve the first time each dev runs the game.

bdach committed:
Remove no longer used resolved storage

bdach committed:
Merge branch 'master' into realm-migration-ui

bdach commented:
I've fixed a code inspection and resolved merge conflicts here. Please let me know if the resolution looks correct.

Otherwise this looks pretty good. One minor detail that it may be good to fix is to move the following log statement:

https://github.com/ppy/osu/blob/93a2a2df35bc88edd4c9db0aabf46f60038307d7/osu.Game/Database/DatabaseContextFactory.cs#L154-L155

out of `DatabaseContextFactory.Backup()` and into the `EFToRealmMigrator` screen, especially so that the two aren't really tied together that tightly (backups run in base game after all). But given that's a dev-only log statement then maybe fine to leave?

peppy commented:
Resolution looks good. Up to you whether you move the log, I don't think it matters too much (it's at the point of the EF database being moved out, so I think it makes sense where it is?)

bdach committed:
Move log statement about migration completed closer to rest of migration code

bdach commented:
> Up to you whether you move the log

Have applied 7e68371.
